Support large vrt

Ryanf55 opened the pull request:
This PR adds support for loading terrain from the ArduPilot terrain server. I've manually generated the VRT dataset that refers to all the tiles. It's a massive dataset, and if you try loading it on `ros2` branch, it will cause a runtime memory allocation failure crash from trying to read 4TB from the GDAL dataset. 

To fix it, I added ROS-configurable limits on the grid map size, and set them to reasonable defaults such that none of the other tif files in your `terrain-models/models/*.tif` have problems.

I've also done some drive-by improvements such as
* Switch from bind to lambda
* Replace tif specific naming to `gdal_dataset` since that's whats more accurately supported
* Improve some error messages
* Add a few assertions to catch errors when run in debug mode
* Add instructions and launch support for running under GDB
* Add const in some places it was missing
* Add ability to just set the path to the DEM file, and then ignore color dataset. Previously, the default was ".", but the validity check was looking for empty string, so no data would be published. 

This needs some additional testing and fixing to better define how to select the origin before it's merged. Right now, I'm just loading the corner of the dataset, but  perhaps this should be a ROS param to specify the lat-lon value for the origin?

Ryanf55 commented on resources/ap_srtm1.vrt:
Ideally, ArduPilot loads this into their terrain server and we use `vsizip` to pull it down. I could zip this file in the repo so it takes up less space. It's 12MB.

Ryanf55 commented:
I attempted to try loading the sargans color tif file at the same time as the SRTM1 dataset for the same area. The pixel resolutions are different, and the data raster sizes are different. The maps don't align, and it causes the grid_map iterator to crash with a floating point exception. I haven't really been able to dig into it yet, but perhaps this would be OK to merge as-is without the capability to handle layers with different sizes/resolutions/datums?

Jaeyoung-Lim commented:
> I saw your open PR to align multiple maps may be related...

Exactly, you would need https://github.com/ethz-asl/grid_map_geo/pull/36 to align multiple DEMs

Jaeyoung-Lim commented on src/map_publisher.cpp:
```diff
@@ -52,18 +54,64 @@ using namespace std::chrono_literals;
 
     map_ = std::make_shared<GridMapGeo>(frame_id);
-    map_->Load(file_path, color_path);
+    map_->setMaxMapSizePixels(max_map_width, max_map_height);
```
What happens if there are no parameters defined? Shouldn't it just use whatever the tif file has?

Ryanf55 replied:
Ah yea, that sounds like better behavior. I'll add that in.

Ryanf55 replied:
I changed the defaults to be infinite width and height, so it just uses what the raster size is.

For loading large rasters like the SRTM VRT, you must configure limits to not have a runtime memory allocation crash.

Jaeyoung-Lim commented on config/sargans_color_over_srtm1.yaml:
```diff
@@ -0,0 +1,8 @@
+        gdal_dataset_color_path: "resources/sargans_color.tif"
+        max_map_width: 3601
+        max_map_height: 3601
+        map_origin_latitude: 47.033333
```
Why would we want to declare this not from the file but externally? Is this maybe because the map alignment was not correct before https://github.com/ethz-asl/grid_map_geo/pull/36 ?

Ryanf55 replied:
I needed to declare it somewhere and wanted to provide a reasonable default. Another approach would be to have them as launch arguments. Once we add capability to dynamically load based on UAV location, then this can be removed.

Jaeyoung-Lim commented on README.md:
````diff
@@ -67,8 +70,44 @@ ros2 launch grid_map_geo load_tif_launch.xml
 ros2 launch grid_map_geo load_tif_launch.xml
 ```
 
+To debug the map publisher in GDB:
````
Nice!

Ryanf55 replied:
Yea, super helpful for chasing the segfaults in GDAL.

Jaeyoung-Lim commented on include/grid_map_geo/transform.hpp:
```diff
@@ -58,4 +67,53 @@ Eigen::Vector3d transformCoordinates(ESPG src_coord, ESPG tgt_coord, const Eigen
+  assert(geoTransform.at(4) == 0);  // assume no rotation
+
+  return {(geoCoords.x() - geoTransform.at(0)) / geoTransform.at(1),
+          (geoCoords.y() - geoTransform.at(3)) / geoTransform.at(5)};
```
I am not sure whether we want to be able to freely choose geoCoords within the `grid_map_geo` framework. At least for me, this additional freedom creates confusion. Wouldn't it be better that we utilize tfs that are clearly defined, so that people doing out of frame stuff can reference the tfs directly?

Ryanf55 replied:
This function is currently used when a user loads a large VRT, and wants to set the map origin in geo coordinates. Could you give me a little more detail on how one can set the map origin using TF?

Jaeyoung-Lim replied:
The map is always placed at (0, 0, 0) and the tf specifies how that frame_id is placed in the datum (e.g. CH1903). This way the map is aligned automatically through tfs, and we don't need to worry where the rviz is looking at.

Ryanf55 replied:
If we are dynamically setting the map from the launch file, and eventually loading tiles dynamically, will (0, 0, 0) be where the first map was loaded? 
If there's a way to do it in TF, I am interested. The use case I'm trying to support here is user-provided origin anywhere on the SRTM dataset in lat-lon coordinates; the origins of the SRTM tiles are not something the users need to care about because GDAL opens datasets it needs under the hood automatically.

Jaeyoung-Lim commented on launch/load_vrt_launch.xml:
```diff
@@ -0,0 +1,19 @@
+    <node pkg="tf2_ros" exec="static_transform_publisher" name="world_map" args="--frame-id world --child-frame-id map"/>
+
+    <node pkg="grid_map_geo" exec="map_publisher" name="map_publisher" output="screen"> 
+	    <param name="gdal_dataset_path" value="/vsizip/$(find-pkg-share grid_map_geo)/resources/ap_srtm1.vrt.zip"/>
```
Does this mean that we need to store the whole VRT file locally?

Ryanf55 replied:
Nope; I should switch this over to use the VRT hosted online and remove this from the repo. Thanks for the reminder.

Ryanf55 replied:
c08463048f8915f76bdc9e78fd75bfee7e950268

Deleted it and switched over the web hosted VRT now.

Jaeyoung-Lim reviewed (changes requested):
Sorry that it took me a while to review this.

Thanks for the amazing effort. I think the fact that we can load VRT into grid_map_geo is fascinating!

I have some comments that might be worth discussing.
- I would prefer if we could utilize tf as the primary mechanism of aligning maps.(at least for now) This way we do not need to worry about alignments or storing the origin position of the tif map. It would just automatically align in rviz.
- The maximum map size needs to be always defined? Could we only use it when specified?
- I am a bit worried that the `map_publisher` will become more and more complicated as we add features. Could we have a simple `tif_loader` that demonstrates how you can use the API, and maybe a standard map publisher that is actually useful for loading VRT .etc
- I am not sure if I understand the VRT loading, since it seems like it is mounted locally onto your system. (under resources?). What would be the path to creating a demonstration of dynamically loading small map tiles as an entity(vehicle) navigates through terrain? Is there a way to directly communicate with the server?

Ryanf55 commented:
Thanks for the review

> Sorry that it took me a while to review this.
> 
> Thanks for the amazing effort. I think the fact that we can load VRT into grid_map_geo is fascinating!
> 
> I have some comments that might be worth discussing.
> 
> * I would prefer if we could utilize tf as the primary mechanism of aligning maps.(at least for now) This way we do not need to worry about alignments or storing the origin position of the tif map. It would just automatically align in rviz.

Yea, this sounds like a good idea. Any recommendations? 

> * The maximum map size needs to be always defined? Could we only use it when specified?

I adjusted the defaults to allow infinite size, but the params for max map size can still be overridden at launch time, which is necessary for the SRTM1 VRT.

> * I am a bit worried that the `map_publisher` will become more and more complicated as we add features. Could we have a simple `tif_loader` that demonstrates how you can use the API, and maybe a standard map publisher that is actually useful for loading VRT .etc

Yes, I can split it up if you like. There will be some code duplication; hope that's ok.

> * I am not sure if I understand the VRT loading, since it seems like it is mounted locally onto your system. (under resources?). What would be the path to creating a demonstration of dynamically loading small map tiles as an entity(vehicle) navigates through terrain? Is there a way to directly communicate with the server?

Fixed, all the examples use the file on the server now.

Jaeyoung-Lim commented:
> > Sorry that it took me a while to review this.
> > Thanks for the amazing effort. I think the fact that we can load VRT into grid_map_geo is fascinating!
> > I have some comments that might be worth discussing.
> > I would prefer if we could utilize tf as the primary mechanism of aligning maps.(at least for now) This way we do not need to worry about alignments or storing the origin position of the tif map. It would just automatically align in rviz.
> Yea, this sounds like a good idea. Any recommendations?

In principle, you should not need to do anything, as long as the tf is published with the correct datum: e.g. https://github.com/ethz-asl/grid_map_geo/blob/02189d61cdc30c384e59a76ba1d044e002c39ba9/src/test_tif_loader.cpp#L76-L90

The only problem is that the SRTM DEMs are represented in WGS84, but ROS coordiantes are orthograpthic(e.g. ENU). We can probably ignore the distortion for now and represent everything in WGS 84(e.g. x=46.0 means 46.0 degrees in longitude). Long term we should probably reproject this into UTM to be useful for actual fusion with other map representations.

Jaeyoung-Lim commented on src/grid_map_geo.cpp:
```diff
@@ -104,41 +139,82 @@ bool GridMapGeo::initializeFromGeotiff(const std::string &path) {
-  double mapcenter_n = originY + pixelSizeY * height * 0.5;
-  maporigin_.espg = ESPG::CH1903_LV03;
-  maporigin_.position = Eigen::Vector3d(mapcenter_e, mapcenter_n, 0.0);
+  if (std::isnan(maporigin_.position.x()) || std::isnan(maporigin_.position.y())) {
```
This would just ignore the actual origin of the map, right? Wouldn't it be better to take whatever is specified in the map?

Ryanf55 replied:
In the SRTM1 data file, the map origin is latitude=0, longitude=0, and it's in the middle of the ocean and there is no data available. So yes, if the user does not specify a reasonable origin over land, then it will try loading tiles over the ocean which are entirely void data.

For now, I was just hoping you can launch in a non-hard-coded location (it's in the ROS params). In a future PR, I was planning to add more intelligent handling of where to launch the origin.

Jaeyoung-Lim replied:
How does it actually work when we are trying query specific regions of the VRT? Shouldn't we get the origin of the parsed datum when the smaller terrain file is downloaded? The way tif loading works is that you never need to specify where the origin is.

Ryanf55 replied:
The advantage of using a large VRT is the actual raster tile sizes and origins become data details; it's up to the user to use application-specifics heuristics for setting the origin. For most UAV applications, basing on the current UAV location should work ok.

Jaeyoung-Lim commented:
@Ryanf55 Any updates regarding this direction?

Ryanf55 commented:
> @Ryanf55 Any updates regarding this direction?



> @Ryanf55 Any updates regarding this direction?

Sorry for the delay. Busy time at work and little time for FOSS. 

I've done a quick rebase and squash. Let me know what you would like with the launch files, and I'll get those cleaned up. The map publisher and tif loader are now separate.

Jaeyoung-Lim reviewed (approved):
LGTM! Thanks!

I think we can iterate on how we want to enable the rolling spawning of maps then we will see what the best API would look like

Jaeyoung-Lim commented on launch/load_vrt_launch.xml:
```diff
@@ -0,0 +1,21 @@
+    <node pkg="tf2_ros" exec="static_transform_publisher" name="world_map" args="--frame-id world --child-frame-id map"/>
+
+    <node pkg="grid_map_geo" exec="map_publisher" name="map_publisher" output="screen"> 
+	    <param name="gdal_dataset_path" value="/vsizip/vsicurl/https://terrain.ardupilot.org/SRTM1/ap_srtm1.zip"/>
```
Shouldn't we keep the tif loader here?

Jaeyoung-Lim commented on launch/load_tif_launch.xml:
```diff
@@ -3,9 +3,13 @@
-    <node pkg="grid_map_geo" exec="test_tif_loader" name="test_tif_loader" namespace="grid_map_geo" output="screen">
-        <param name="tif_path" value="$(find-pkg-share grid_map_geo)/resources/sargans.tif"/>
-        <param name="tif_color_path" value="$(find-pkg-share grid_map_geo)/resources/sargans_color.tif"/>
+    <node pkg="grid_map_geo" exec="map_publisher" name="map_publisher" namespace="grid_map_geo" output="screen">
```
Shouldn't we keep the tif loader here?

Ryanf55 replied:
Yea. I was curious why have both python and XML launch files for TIF, and do you want both for the VRT loader?

Jaeyoung-Lim replied:
No real preference :) whatever is more convenient

Ryanf55 replied:
Fixed.

Ryanf55 commented:
Remaining tasks I'd like before merge if you aren't in a rush
- [ ] Write tests for all the launch files, disable RVIZ, and make sure the nodes don't crash, and make sure they publish a non-empty map
- [ ] Fix any bugs from testing (I think the original tif loader doesn't work because we need to set the origin to the raster origin.

Jaeyoung-Lim commented:
@Ryanf55 I have started looking into implementing a rolling query depending on vehicle position. 

What I learned in between though is that ROS 2 is not really usable in the field over LTE with large messages (elevation maps/images). It seems to be a known issue without a clear solution. Something we need to keep in mind before people use this package in the field

Jaeyoung-Lim commented on config/sargans_color_over_srtm1.yaml:
```diff
@@ -0,0 +1,8 @@
+map_publisher:
+    ros__parameters:
+        gdal_dataset_path: /vsizip/vsicurl/https://terrain.ardupilot.org/SRTM1/N09E047.hgt.zip/N09E047.hgt
```
Where does the `/vsizip/vsicurl` come from?

I tried to directly query the terrain server with gdalinfo, but didn't work:
```
$ gdalinfo https://terrain.ardupilot.org/SRTM1/N09E047.hgt.zip/N09E047.hgt
ERROR 1: An error occurred while creating a virtual connection to the DAP server:Error while reading the URL: https://terrain.ardupilot.org/SRTM1/N09E047.hgt.zip/N09E047.hgt.ver.
The OPeNDAP server returned the following message:
Not Found: The data source or server could not be found.
        Often this means that the OPeNDAP server is missing or needs attention.
        Please contact the server administrator.
gdalinfo failed - unable to open 'https://terrain.ardupilot.org/SRTM1/N09E047.hgt.zip/N09E047.hgt'.

```

Jaeyoung-Lim commented:
Since elevation unit is in `m`, and the x, y in degrees for WGS84 this is what it looks like when visualized in Rviz:

![rviz_screenshot_2024_04_08-17_19_33](https://github.com/ethz-asl/grid_map_geo/assets/5248102/c4c5bc5e-bf40-44b6-8bc9-e873de35e641)
